COLLABORATOR_GUIDE.md

```diff
@@ -391,6 +391,27 @@ Runtime Deprecations and End-of-life APIs (internal or public) must be
 handled as semver-major changes unless there is TSC consensus to land the
 deprecation as a semver-minor.
 
+Possible reasons for Runtime deprecations can be:
+
+* Node.js will no longer be able to support the API, for example if required
+  code is being removed in an upstream project.
+
+* The API is too hard or impossible to use correctly, increasing the risk of
+  undetected bugs or security issues in existing code.
+
+* Breaking the API makes way for simplification of its implementation that is
+  significant enough to make maintainability of Node.js easier, or enables
+  implementation of features that would otherwise be impossible.
+
+In addition, a Documentation-Only deprecation may be issued if it is desirable
+to recommend a single canonical way out of multiple different ones
+which achieve a particular goal with Node.js.
+
+In particular, a Runtime deprecation should *not* be added to an API that
+that requires only trivial maintenance, or only because it has
+never been documented or has only accidentally been exposed and there is
+usage of said API in existing ecosystem code.
```

Comment:
I disagree that a runtime deprecation should not be used in case an API has never been documented before. Especially in such a case I would want a runtime deprecation as otherwise it is highly unlikely that anyone will ever realize that a specific code is actually deprecated and was never documented.

Also accidentally exposing a API to the ecosystem should receive a runtime deprecation out of my perspective.

Reply:
> otherwise it is highly unlikely that anyone will ever realize that a specific code is actually deprecated and was never documented.

I fail to see how that is a bad thing?

Reply:
> I fail to see how that is a bad thing?

I am closer to the viewpoint that @addaleax has on this, i.e. this is not necessarily a bad thing. However, one has to recognize that we end up with a large number of documentation-only deprecated stuff (i.e. bloat). That's the downside, and I personally think it is okay. I don't think we are too bloated ATM.

Reply:
If we add this, I would like a note that runtime deprecation may occur if a previously undocumented or accidentally exposed API is given a documented and intentional API.

Reply:
I would agree with @bmecks' caveat here. If there is a supported replacement, we ought to be able to runtime deprecate the trivial unsupported thing.

Reply:
Everything else looks good tho

Reply:
I'm not ok with this change, at least the way I read this. We keep stumbling on non-documented APIs used in the ecosystem that breaks their codebase on a new node major. We need to be able to runtime-deprecate them easily.

I think we should be able to runtime-deprecate them even if we do not provide an alternative.

Reply:
> We need to be able to runtime-deprecate them easily

'We need to' is a strong phrase. Why do we *need* do? I think it is cost vs. benefit. The premise here is already APIs that are trivial to maintain. I see bloat accumulating as cost, but that shouldn't drive a *need* to runtime deprecated specially if there is ecosystem impact.

Reply:
> We keep stumbling on non-documented APIs used in the ecosystem that breaks their codebase on a new node major. We need to be able to runtime-deprecate them easily.

Just to be clear, not runtime-deprecating would imply not removing/breaking them.

> I think we should be able to runtime-deprecate them even if we do not provide an alternative.

@mcollina What I'm saying is that "they are not documented" is not a sufficient reason for removal. Are you disagreeing with that? Else I'm not sure what you're saying here...

> I would like a note that runtime deprecation may occur if a previously undocumented or accidentally exposed API is given a documented and intentional API.

@bmeck @jasnell I have mixed feelings about that... if it's basically a situation in which we expose previously undocumented API in a well-defined public manner, what's the cost of keeping an alias indefinitely? If it's a more complex situation in which the new API differs substantially from the undocumented one, isn't that covered by the listed reasons for runtime deprecations?

Reply:
We need to be able is not that strong, it leaves the door open. The sentence in this change says we should *not* do it. I see every non-documented public API as a possible source of future breakage in the ecosystem. I would prefer those non-documented APIs to be deprecated, and add the relevant public APIs to keep solving the same use case. If a non-documented API is not very widespread in the ecosystem, it's the right moment to runtime deprecate it (and potentially replace with a new one), because it has a chance to become popular.

(@bmeck proposal is a good middle-ground solution that I can stand behind).

Reply:
> and there is usage

There is noticeable/significant usage?
I don't think a single abadoned line of code on GitHub would justify keeping the API.

Reply:
I think one reason for the different point of views here are what we understand with a breaking change and more importantly: how "bad" a runtime deprecation is for each of us. For me a runtime deprecation means there is just something logged so people get notified that it is not recommended to use a specific API anymore / in a specific way. It does not even mean that something might or might not be removed. I feel like that is nothing bad - almost ever.

Without that notification we reach probably less than a thousandths of the users and probably especially those for whom the deprecation is actually meant for. A documentation deprecation is only useful out of my perspective when we want to only prevent new people from using something. But even that is often probably not a reason for people to not use something, especially as a lot of people do not read a documentation properly and they might just check a online example that they found somewhere else than in the official documentation.
Instead we are able to shift code to become better and to outline that things are not ideal and I think that is something great.

So I feel like we should always have the possibility of runtime deprecating things that were never meant to be exposed or that were never documented. In some cases it might be useful to add a official API instead, on other cases there might be nothing to do besides deprecating it to tell people to better switch and in even more cases there are probably other solutions.

Documenting something and than runtime deprecating is the opposite of what I would see as logical because a lot more people will actually read the documentation about the deprecation (especially new people) than people who try to use a API they stumble upon somewhere in the Node.js source code. So for me it is like the opposite step of what is meant to be achieved.

Deciding on what to do should out of my perspective be individual per case. I do not want to have a strict limitation one way or the other.

Reply:
> I would prefer those non-documented APIs to be deprecated, and add the relevant public APIs to keep solving the same use case.

I think nobody's disagreeing with that. :)

> If a non-documented API is not very widespread in the ecosystem, it's the right moment to runtime deprecate it (and potentially replace with a new one), because it has a chance to become popular.

Okay, I can see that. Let me think about it.

> For me a runtime deprecation means there is just something logged so people get notified that it is not recommended to use a specific API anymore / in a specific way. 

The feedback that we've gotten from the Buffer deprecation discussions is different: A runtime deprecation *is* functional breakage in itself.

> A documentation deprecation is only useful out of my perspective when we want to only prevent new people from using something.

I agree.

> Documenting something and than runtime deprecating is the opposite of what I would see as logical because a lot more people will actually read the documentation about the deprecation (especially new people) than people who try to use a API they stumble upon somewhere in the Node.js source code. 

Please note that that's not what this PR is suggesting; I am totally fine with considering an undocumented API to be implicitly docs-deprecated for the purposes of this policy.

Reply:
> Please note that that's not what this PR is suggesting; I am totally fine with considering an undocumented API to be implicitly docs-deprecated for the purposes of this policy.

I just mentioned that because it was brought up.

> The feedback that we've gotten from the Buffer deprecation discussions is different: A runtime deprecation is functional breakage in itself.

Well, I personally think that is a strange view. The code itself still works perfectly fine, so having a warning is something that I would actually want. And especially the Buffer deprecation is something where there are a lot of different opinions.

Reply:
This paragraph contains one sentence that seems overly long and perhaps difficult to understand. Consider breaking it into multiple sentences. Perhaps something like this?:

```text
A Runtime deprecation should not be added to an API that
requires only trivial maintenance. A Runtime deprecation should
not be added simply because an API has never been documented
(or has only accidentally been exposed) if there is significant usage
of the API in the ecosystem.
```